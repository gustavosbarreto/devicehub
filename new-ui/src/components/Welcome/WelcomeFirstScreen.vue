<template>
  <p class="ml-4 pt-4 text-subtitle-2">
    Hello, <strong> {{ name }} </strong> !
  </p>
  <p class="mt-4 ml-4 mr-4 text-subtitle-2">
    ShellHub is a modern SSH server for remotely accessing Linux devices via CLI
    or web-based user interface, it avoids the inconveniences such as firewall
    and NAT when getting your IP address, automating the access process, and
    helping you easily, seamlessly and securely connect your device to the
    internet.
  </p>
</template>

<script lang="ts">
import { defineComponent, computed } from "vue";
import { useStore } from "../../store";

export default defineComponent({
  setup() {
    const store = useStore();

    const name = computed(
      () => store.getters["auth/currentName"] || store.getters["auth/currentUser"],
    );
    return {
      name,
    };
  },
});
</script>
