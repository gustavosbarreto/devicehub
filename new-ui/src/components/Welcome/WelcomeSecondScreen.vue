<template>
  <p class="ml-4 pt-4 text-subtitle-2">
    In order to register a device on ShellHub, you need to install ShellHub
    agent onto it.
  </p>
  <p class="ml-4 pt-4 text-subtitle-2">
    The easiest way to install ShellHub agent is with our automatic one-line
    installation script, which works with all Linux distributions that have
    Docker installed and properly set up.
  </p>
  <div class="mt-4 ml-4 mr-4">
    <p class="ml-2 pt-4 text-subtitle-2 text-bold">
      Run the following command on your device:
    </p>

    <v-text-field
      :model-value="command"
      @click:append="copyCommand"
      class="code mt-1"
      variant="outlined"
      append-icon="mdi-content-copy"
      readonly
      active
      data-test="command-field"
      density="compact"
    />

    <v-divider />

    <p class="text-caption mb-0 mt-1">
      Check the
      <a
        :href="'https://docs.shellhub.io/user-guides/devices/adding'"
        target="_blank"
        >documentation</a
      >
      for more information and alternative install methods.
    </p>
  </div>
</template>

<script lang="ts">
import { INotificationsCopy } from "../../interfaces/INotifications";
import { useStore } from "../../store";
import { defineComponent } from "vue";

export default defineComponent({
  props: {
    command: {
      type: String,
      required: true,
    },
  },

  setup(props) {
    const store = useStore();
    const copyCommand = () => {
      navigator.clipboard.writeText(props.command);
      store.dispatch("snackbar/showSnackbarCopy", INotificationsCopy.tenantId);
    };

    return {
      copyCommand,
    };
  },
});
</script>
