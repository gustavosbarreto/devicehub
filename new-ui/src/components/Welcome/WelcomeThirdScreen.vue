<template>
  <div class="pa-4">
    <p class="ml-4 pt-4 text-subtitle-2">
      A device connection has been detected.
    </p>
    <p class="ml-4 pt-4 text-subtitle-2">
      Please confirm that this device is yours to enroll into your account.
      After confirmation, you will go to the last step of introducing
      <strong>ShellHub</strong>.
    </p>

    <v-row no-gutters class="mt-4">
      <v-col>
        <v-card class="pa-2" tile :elevation="0">
          <strong>Hostname</strong>
        </v-card>
        <v-card
          class="pa-2"
          tile
          :elevation="0"
          data-test="deviceName-field"
        >
          {{ getPendingDevice.name }}
        </v-card>
      </v-col>
      <v-col>
        <v-card class="pa-2" tile :elevation="0">
          <strong>Operation System</strong>
        </v-card>
        <v-card
          class="pa-2"
          tile
          :elevation="0"
          data-test="devicePrettyName-field"
        >
          <div v-if="getPendingDevice.info">
            <DeviceIcon :icon="getPendingDevice.info.id" />
            {{ getPendingDevice.info.pretty_name }}
          </div>
        </v-card>
      </v-col>
    </v-row>
  </div>
</template>

<script lang="ts">
import { defineComponent, computed, onMounted } from "vue";
import { useStore } from "../../store";
import { INotificationsError } from "../../interfaces/INotifications";
import DeviceIcon from "../Devices/DeviceIcon.vue";

export default defineComponent({
  setup() {
    const store = useStore();
    const getPendingDevice = computed(
      () => store.getters["devices/getFirstPending"],
    );
    onMounted(() => {
      try {
        store.dispatch("devices/setFirstPending");
      } catch {
        store.dispatch(
          "snackbar/showSnackbarErrorLoading",
          INotificationsError.devicePending,
        );
      }
    });
    return {
      getPendingDevice,
    };
  },
  components: { DeviceIcon },
});
</script>
