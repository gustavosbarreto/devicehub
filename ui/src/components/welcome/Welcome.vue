<template>
  <v-card>
    <v-dialog
      v-model="showWelcome"
      :retain-focus="false"
      max-width="800px"
      persistent
    >
      <v-stepper v-model="e1">
        <v-stepper-header>
          <v-stepper-step
            :complete="e1 > 1"
            step="1"
          >
            Welcome
          </v-stepper-step>

          <v-divider />

          <v-stepper-step
            :complete="e1 > 2"
            step="2"
          >
            Connecting device
          </v-stepper-step>

          <v-divider />

          <v-stepper-step
            :complete="e1 > 3"
            step="3"
          >
            Authorizing device
          </v-stepper-step>

          <v-divider />

          <v-stepper-step step="4">
            Finish
          </v-stepper-step>
        </v-stepper-header>

        <v-stepper-items>
          <v-stepper-content step="1">
            <v-card
              class="mb-12"
              height="250px"
            >
              <WelcomeFirstScreen />
            </v-card>
            <v-card-actions>
              <v-btn
                text
                @click="close"
              >
                Close
              </v-btn>
              <v-spacer />
              <v-btn
                color="primary"
                data-test="firstClick-btn"
                @click="activePollingDevices()"
              >
                Next
              </v-btn>
            </v-card-actions>
          </v-stepper-content>

          <v-stepper-content step="2">
            <v-card
              class="mb-12"
              height="250px"
            >
              <WelcomeSecondScreen
                :command="command()"
              />
            </v-card>
            <v-card-actions>
              <v-btn
                text
                data-test="close-btn"
                @click="close"
              >
                Close
              </v-btn>
              <v-spacer />
              <v-btn
                @click="e1 = 1"
              >
                Back
              </v-btn>
              <v-btn
                v-if="!enable"
                disabled
              >
                Waiting for Device
              </v-btn>
              <v-btn
                v-else
                color="primary"
                data-test="secondClick-btn"
                @click="e1 = 3"
              >
                Next
              </v-btn>
            </v-card-actions>
          </v-stepper-content>

          <v-stepper-content step="3">
            <v-card
              class="mb-12"
              height="250px"
            >
              <WelcomeThirdScreen
                v-if="enable"
              />
            </v-card>
            <v-card-actions>
              <v-btn
                text
                @click="close"
              >
                Close
              </v-btn>
              <v-spacer />
              <v-btn
                @click="e1 = 2"
              >
                Back
              </v-btn>
              <v-btn
                color="primary"
                data-test="thirdClick-btn"
                @click="acceptDevice()"
              >
                Accept
              </v-btn>
            </v-card-actions>
          </v-stepper-content>

          <v-stepper-content step="4">
            <v-card
              class="mb-12"
              height="250px"
            >
              <WelcomeFourthScreen />
            </v-card>
            <v-card-actions>
              <v-spacer />
              <v-btn
                color="primary"
                @click="close"
              >
                Finish
              </v-btn>
            </v-card-actions>
          </v-stepper-content>
        </v-stepper-items>
      </v-stepper>
    </v-dialog>
  </v-card>
</template>

<script>

import WelcomeFirstScreen from '@/components/welcome/WelcomeFirstScreen';
import WelcomeSecondScreen from '@/components/welcome/WelcomeSecondScreen';
import WelcomeThirdScreen from '@/components/welcome/WelcomeThirdScreen';
import WelcomeFourthScreen from '@/components/welcome/WelcomeFourthScreen';

export default {
  name: 'WelcomeComponent',

  components: {
    WelcomeFirstScreen,
    WelcomeSecondScreen,
    WelcomeThirdScreen,
    WelcomeFourthScreen,
  },

  props: {
    show: {
      type: Boolean,
      required: true,
    },
  },

  data() {
    return {
      e1: 1,
      enable: false,
      polling: null,
      curl: {
        hostname: window.location.hostname,
        tenant: this.$store.getters['auth/tenant'],
      },
    };
  },

  computed: {
    showWelcome: {
      get() {
        return this.show;
      },

      set(value) {
        this.$emit('show', value);
      },
    },
  },

  methods: {
    command() {
      return `curl -sSf "${window.location.protocol}//${this.curl.hostname}/install.sh?tenant_id=${this.curl.tenant}" | sh`;
    },

    activePollingDevices() {
      this.e1 = 2;
      this.pollingDevices();
    },

    pollingDevices() {
      this.polling = setInterval(async () => {
        try {
          await this.$store.dispatch('stats/get');

          this.enable = this.$store.getters['stats/stats'].pending_devices !== 0;
          if (this.enable) {
            this.e1 = 3;
            clearTimeout(this.polling);
          }
        } catch {
          this.$store.dispatch('snackbar/showSnackbarErrorDefault');
        }
      }, 3000);
    },

    async acceptDevice() {
      const device = this.$store.getters['devices/getFirstPending'];
      try {
        await this.$store.dispatch('devices/accept', device.uid);

        this.$store.dispatch('notifications/fetch');
        this.$store.dispatch('stats/get');

        this.e1 = 4;
      } catch {
        this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.deviceAccepting);
      }
    },

    async close() {
      this.$emit('update:show', false);
      clearTimeout(this.polling);
    },
  },
};
</script>
