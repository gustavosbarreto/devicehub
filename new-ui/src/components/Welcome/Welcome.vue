<template>
  <v-dialog
    v-model="showWelcome"
    max-width="800px"
    min-width="60vw"
    :retain-focus="false"
    persistent
  >
    <v-card class="pa-6 bg-grey-darken-4 bg-v-theme-surface">
      <v-card-title class="text-center mb-4">
        <span>Step</span>
        {{ el }}
        <span>of</span>
        4
        <v-divider class="mt-2" />
      </v-card-title>
      <v-window v-model="el">
        <v-window-item :value="1">
          <v-card class="bg-v-theme-surface" height="250px">
            <WelcomeFirstScreen />
          </v-card>
          <v-card-actions class="mt-4">
            <v-btn text @click="close">Close</v-btn>
            <v-spacer />
            <v-btn
              data-test="firstClick-btn"
              color="primary"
              @click="activePollingDevices()"
              >Next</v-btn
            >
          </v-card-actions>
        </v-window-item>

        <v-window-item :value="2">
          <v-card class="bg-v-theme-surface" height="250px">
            <WelcomeSecondScreen :command="command()" />
          </v-card>
          <v-card-actions>
            <v-btn text data-test="close-btn" @click="close"> Close </v-btn>
            <v-spacer />
            <v-btn text @click="el--">back</v-btn>
            <v-btn v-if="!enable" disabled> Waiting for Device </v-btn>
            <v-btn
              v-else
              color="primary"
              data-test="secondClick-btn"
              @click="el = 3"
            >
              Next
            </v-btn>
          </v-card-actions>
        </v-window-item>

        <v-window-item :value="3">
          <v-card class="bg-v-theme-surface" height="250px">
            <WelcomeThirdScreen v-if="enable" />
          </v-card>
          <v-card-actions>
            <v-btn variant="text" data-test="close-btn" @click="close">
              Close
            </v-btn>
            <v-spacer />
            <v-btn variant="text" @click="el--">back</v-btn>
            <v-btn
              color="primary"
              data-test="accept-btn"
              @click="acceptDevice()"
            >
              Accept
            </v-btn>
          </v-card-actions>
        </v-window-item>

        <v-window-item :value="4">
          <v-card class="bg-v-theme-surface" height="250px">
            <WelcomeFourthScreen />
          </v-card>
          <v-card-actions>
            <v-spacer />
            <v-btn color="primary" data-test="finish-btn" @click="close">
              Finish
            </v-btn>
          </v-card-actions>
        </v-window-item>
      </v-window>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import { INotificationsError } from "../../interfaces/INotifications";
import { computed, defineComponent, ref } from "vue";
import { useStore } from "../../store";
import WelcomeFirstScreen from "./WelcomeFirstScreen.vue";
import WelcomeSecondScreen from "./WelcomeSecondScreen.vue";
import WelcomeThirdScreen from "./WelcomeThirdScreen.vue";
import WelcomeFourthScreen from "./WelcomeFourthScreen.vue";

export default defineComponent({
  props: {
    show: {
      type: Boolean,
      required: true,
    },
  },
  emits: ["update"],
  setup(props, ctx) {
    const store = useStore();
    const el = ref<number>(1);
    const polling = ref<any>(null);
    const enable = ref(false);
    const showWelcome = computed({
      get() {
        return props.show;
      },
      set(value: boolean) {
        ctx.emit("update", value);
      },
    });

    const curl = ref({
      hostname: window.location.hostname,
      tenant: store.getters["auth/tenant"],
    });

    const activePollingDevices = () => {
      el.value = 2;
      pollingDevices();
    };

    const pollingDevices = () => {
      polling.value = setInterval(async () => {
        try {
          await store.dispatch("stats/get");

          enable.value = store.getters["stats/stats"].pending_devices !== 0;
          if (enable.value) {
            el.value = 3;
            clearTimeout(polling.value);
          }
        } catch (error: any) {
          store.dispatch("snackbar/showSnackbarErrorDefault");
          throw new Error(error);
        }
      }, 3000);
    };

    const acceptDevice = async () => {
      const device = store.getters["devices/getFirstPending"];
      try {
        await store.dispatch("devices/accept", device.uid);

        store.dispatch("notifications/fetch");
        store.dispatch("stats/get");

        el.value = 4;
      } catch (error: any) {
        store.dispatch(
          "snackbar/showSnackbarErrorAction",
          INotificationsError.deviceAccepting
        );
        throw new Error(error);
      }
    };

    const command = () => {
      let port = window.location.port ? `:${window.location.port}` : "";
      let hostname = window.location.hostname;

      return `curl -sSf "${window.location.protocol}//${hostname}${port}/install.sh?tenant_id=${curl.value.tenant}" | sh`;
    };

    const close = () => {
      ctx.emit("update", false);
      showWelcome.value = false;
      clearTimeout(polling.value);
    };

    return {
      el,
      showWelcome,
      enable,
      polling,
      curl,
      close,
      activePollingDevices,
      acceptDevice,
      command,
    };
  },
  components: {
    WelcomeFirstScreen,
    WelcomeSecondScreen,
    WelcomeThirdScreen,
    WelcomeFourthScreen,
  },
});
</script>
