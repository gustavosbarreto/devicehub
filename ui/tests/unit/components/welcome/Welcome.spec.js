import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Welcome from '@/components/welcome/Welcome';

describe('Welcome', () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);

  let wrapper;

  const show = true;
  const tenant = 'a582b47a42e';
  const stats = {
    registered_devices: 0,
    online_devices: 0,
    active_sessions: 0,
    pending_devices: 1,
    rejected_devices: 0,
  };
  const devicePending = {
    uid: 'a582b47a42e',
    name: '39-5e-2b',
    identity: {
      mac: '00:00:00:00:00:00',
    },
    info: {
      id: 'linuxmint',
      pretty_name: 'Linux Mint 19.3',
      version: '',
    },
    public_key: '----- PUBLIC KEY -----',
    tenant_id: '00000001',
    last_seen: '2020-05-20T19:58:53.276Z',
    online: true,
    namespace: 'user',
    status: 'accepted',
  };

  const store = new Vuex.Store({
    namespaced: true,
    state: {
      tenant,
      stats,
      devicePending,
    },
    getters: {
      'auth/tenant': (state) => state.tenant,
      'stats/stats': (state) => state.stats,
      'devices/getFirstPending': (state) => state.devicePending,
    },
    actions: {
      'stats/get': () => {
      },
      'devices/accept': () => {
      },
      'notifications/fetch': () => {
      },
      'modals/showSnackbarErrorAction': () => {
      },
      'modals/showSnackbarErrorDefault': () => {
      },
    },
  });

  beforeEach(() => {
    wrapper = shallowMount(Welcome, {
      store,
      localVue,
      stubs: ['fragment'],
      propsData: { show },
    });
  });

  it('Is a Vue instance', () => {
    expect(wrapper).toBeTruthy();
  });
  it('Renders the component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('Receive data in props', () => {
    expect(wrapper.vm.show).toEqual(show);
  });
  it('Process data in the methods', () => {
    const command = `curl "http://localhost/install.sh?tenant_id=${tenant}" | sh`;
    expect(wrapper.vm.command()).toEqual(command);
  });
  it('Compare data with default value', () => {
    expect(wrapper.vm.e1).toEqual(1);
    expect(wrapper.vm.enable).toEqual(false);
    expect(wrapper.vm.curl.hostname).toEqual('localhost');
    expect(wrapper.vm.curl.tenant).toEqual(tenant);
  });
  it('Test the click event', async () => {
    wrapper.find('[data-test="firstClick-btn"]').vm.$emit('click');
    expect(wrapper.vm.e1).toEqual(2);

    wrapper.setData({ enable: true });
    await localVue.nextTick();
    wrapper.find('[data-test="secondClick-btn"]').vm.$emit('click');
    expect(wrapper.vm.e1).toEqual(3);

    await localVue.nextTick();
    expect(wrapper.find('[data-test="thirdClick-btn"]').exists()).toBe(true);
    await wrapper.vm.acceptDevice();
    expect(wrapper.vm.e1).toEqual(4);
  });
});
