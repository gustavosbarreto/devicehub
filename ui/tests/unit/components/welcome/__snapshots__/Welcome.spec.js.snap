// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome Renders the component 1`] = `
<v-card-stub loaderheight="4" tag="div">
  <v-dialog-stub opendelay="0" closedelay="0" value="true" openonclick="true" contentclass="" maxwidth="800px" origin="center center" persistent="true" transition="dialog-transition">
    <v-stepper-stub tag="div" value="1">
      <v-stepper-header-stub tag="div">
        <v-stepper-step-stub color="primary" completeicon="$complete" editicon="$edit" erroricon="$error" rules="" step="1">
          Welcome
        </v-stepper-step-stub>
        <v-divider-stub></v-divider-stub>
        <v-stepper-step-stub color="primary" completeicon="$complete" editicon="$edit" erroricon="$error" rules="" step="2">
          Connecting device
        </v-stepper-step-stub>
        <v-divider-stub></v-divider-stub>
        <v-stepper-step-stub color="primary" completeicon="$complete" editicon="$edit" erroricon="$error" rules="" step="3">
          Authorizing device
        </v-stepper-step-stub>
        <v-divider-stub></v-divider-stub>
        <v-stepper-step-stub color="primary" completeicon="$complete" editicon="$edit" erroricon="$error" rules="" step="4">
          Finish
        </v-stepper-step-stub>
      </v-stepper-header-stub>
      <v-stepper-items-stub tag="div">
        <v-stepper-content-stub step="1">
          <v-card-stub loaderheight="4" tag="div" height="250px" class="mb-12">
            <welcomefirstscreen-stub></welcomefirstscreen-stub>
          </v-card-stub>
          <v-card-actions-stub tag="div">
            <v-btn-stub tag="button" activeclass="" text="true" type="button">
              Close
            </v-btn-stub>
            <v-spacer-stub tag="div"></v-spacer-stub>
            <v-btn-stub color="primary" tag="button" activeclass="" type="button" data-test="firstClick-btn">
              Next
            </v-btn-stub>
          </v-card-actions-stub>
        </v-stepper-content-stub>
        <v-stepper-content-stub step="2">
          <v-card-stub loaderheight="4" tag="div" height="250px" class="mb-12">
            <welcomesecondscreen-stub command="curl -sSf &quot;http://localhost/install.sh?tenant_id=a582b47a42e&quot; | sh"></welcomesecondscreen-stub>
          </v-card-stub>
          <v-card-actions-stub tag="div">
            <v-btn-stub tag="button" activeclass="" text="true" type="button" data-test="close-btn">
              Close
            </v-btn-stub>
            <v-spacer-stub tag="div"></v-spacer-stub>
            <v-btn-stub tag="button" activeclass="" type="button">
              Back
            </v-btn-stub>
            <v-btn-stub tag="button" activeclass="" disabled="true" type="button">
              Waiting for Device
            </v-btn-stub>
          </v-card-actions-stub>
        </v-stepper-content-stub>
        <v-stepper-content-stub step="3">
          <v-card-stub loaderheight="4" tag="div" height="250px" class="mb-12">
            <!---->
          </v-card-stub>
          <v-card-actions-stub tag="div">
            <v-btn-stub tag="button" activeclass="" text="true" type="button">
              Close
            </v-btn-stub>
            <v-spacer-stub tag="div"></v-spacer-stub>
            <v-btn-stub tag="button" activeclass="" type="button">
              Back
            </v-btn-stub>
            <v-btn-stub color="primary" tag="button" activeclass="" type="button" data-test="thirdClick-btn">
              Accept
            </v-btn-stub>
          </v-card-actions-stub>
        </v-stepper-content-stub>
        <v-stepper-content-stub step="4">
          <v-card-stub loaderheight="4" tag="div" height="250px" class="mb-12">
            <welcomefourthscreen-stub></welcomefourthscreen-stub>
          </v-card-stub>
          <v-card-actions-stub tag="div">
            <v-spacer-stub tag="div"></v-spacer-stub>
            <v-btn-stub color="primary" tag="button" activeclass="" type="button">
              Finish
            </v-btn-stub>
          </v-card-actions-stub>
        </v-stepper-content-stub>
      </v-stepper-items-stub>
    </v-stepper-stub>
  </v-dialog-stub>
</v-card-stub>
`;
